db/write_callback_test.cc

```diff
@@ -124,254 +149,238 @@ TEST_F(WriteCallbackTest, WriteWithCallbackTest) {
-                }
-
-                WriteOptions woptions;
-                woptions.disableWAL = !enable_WAL;
-                woptions.sync = enable_WAL;
-                Status s;
-                if (seq_per_batch) {
-                  class PublishSeqCallback : public PreReleaseCallback {
-                   public:
-                    PublishSeqCallback(DBImpl* db_impl_in)
-                        : db_impl_(db_impl_in) {}
-                    Status Callback(SequenceNumber last_seq, bool /*not used*/,
-                                    uint64_t, size_t /*index*/,
-                                    size_t /*total*/) override {
-                      db_impl_->SetLastPublishedSequence(last_seq);
-                      return Status::OK();
-                    }
-                    DBImpl* db_impl_;
-                  } publish_seq_callback(db_impl);
-                  // seq_per_batch requires a natural batch separator or Noop
-                  WriteBatchInternal::InsertNoop(&write_op.write_batch_);
-                  const size_t ONE_BATCH = 1;
-                  s = db_impl->WriteImpl(
-                      woptions, &write_op.write_batch_, &write_op.callback_,
-                      nullptr, 0, false, nullptr, ONE_BATCH,
-                      two_queues ? &publish_seq_callback : nullptr);
-                } else {
-                  s = db_impl->WriteWithCallback(
-                      woptions, &write_op.write_batch_, &write_op.callback_);
-                }
-
-                if (write_op.callback_.should_fail_) {
-                  ASSERT_TRUE(s.IsBusy());
-                } else {
-                  ASSERT_OK(s);
-                }
-              };
-
-              ROCKSDB_NAMESPACE::SyncPoint::GetInstance()->EnableProcessing();
-
-              // do all the writes
-              std::vector<port::Thread> threads;
-              for (uint32_t i = 0; i < write_group.size(); i++) {
-                threads.emplace_back(write_with_callback_func);
-              }
-              for (auto& t : threads) {
-                t.join();
-              }
-
-              ROCKSDB_NAMESPACE::SyncPoint::GetInstance()->DisableProcessing();
-
-              // check for keys
-              string value;
-              for (auto& w : write_group) {
-                ASSERT_TRUE(w.callback_.was_called_.load());
-                for (auto& kvp : w.kvs_) {
-                  if (w.callback_.should_fail_) {
-                    ASSERT_TRUE(
-                        db->Get(read_options, kvp.first, &value).IsNotFound());
-                  } else {
-                    ASSERT_OK(db->Get(read_options, kvp.first, &value));
-                    ASSERT_EQ(value, kvp.second);
-                  }
-                }
-              }
-
-              ASSERT_EQ(seq.load(), db_impl->TEST_GetLastVisibleSequence());
-
-              delete db;
-              DestroyDB(dbname, options);
-            }
+  for (auto& write_group : write_scenarios) {
```

Comment:
These scenarios could also be tested independently. The way to do it is to introduce a static helper `GetWriteScenarios` that returns the above `std::vector<std::vector<WriteOP>>` and add `testing::ValuesIn(GetWriteScenarios())` to the `testing::Combine` call below.

Reply:
Maybe leave to next TechDebt week.